cirq-google/cirq_google/study/zip_longest.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Dict, Iterator, List
+
+import itertools
+import cirq
+
+
+class ZipLongest(cirq.Zip):
+    """Iterate over constituent sweeps in parallel
+
+    Analogous to itertools.zip_longest.
+    Note that we iterate until all sweeps terminate,
+    so if the sweeps are different lengths, the
+    shorter sweeps will be filled by repeating their last value
+    until all sweeps have equal length.
+    This is different from itertools.zip_longest, which uses a fixed fill value.
+    """
+
+    def __init__(self, *sweeps: cirq.Sweep) -> None:
+        self.sweeps = sweeps
+
+    def __eq__(self, other):
+        if not isinstance(other, ZipLongest):
+            return NotImplemented
+        return self.sweeps == other.sweeps
+
+    def __hash__(self) -> int:
+        return hash(tuple(self.sweeps))
+
+    @property
+    def keys(self) -> List['cirq.TParamKey']:
+        return sum((sweep.keys for sweep in self.sweeps), [])
```

Comment:
These are the same as in the base class, can we drop them?

Reply:
Dropped keys().  I am not sure why, but it seems to need the magic dunder function hash in the subclass.